driver-embedded/src/main/com/mongodb/embedded/client/MongoDBCAPIHelper.java

```diff
@@ -64,108 +67,103 @@ static synchronized void init(final MongoEmbeddedSettings mongoEmbeddedSettings)
+            status = null;
         } catch (Throwable t) {
-            throw createError("db_new", t);
+            throw createError("status_destroy", t);
         }
+
+        mongoDBCAPI = null;
     }
 
-    static void db_destroy(final Pointer db) {
+    static Pointer instance_create(final String yamlConfig) {
         checkInitialized();
         try {
-            mongoDBCAPI.libmongodbcapi_db_destroy(db);
+            return validatePointerCreated(mongoDBCAPI.libmongodbcapi_instance_create(lib, yamlConfig, status));
         } catch (Throwable t) {
-            throw createError("db_destroy", t);
+            throw createError("instance_create", t);
         }
     }
 
-    static void db_pump(final Pointer db) {
+    static void instance_destroy(final Pointer instance) {
         checkInitialized();
         try {
-            validateErrorCode(mongoDBCAPI.libmongodbcapi_db_pump(db));
+            validateErrorCode(mongoDBCAPI.libmongodbcapi_instance_destroy(instance, status));
         } catch (Throwable t) {
-            throw createError("db_pump", t);
+            throw createError("instance_destroy", t);
         }
     }
 
-    static Pointer db_client_new(final Pointer db) {
+    static Pointer create_client(final Pointer instance) {
         checkInitialized();
         try {
-            return mongoDBCAPI.libmongodbcapi_db_client_new(db);
+            return validatePointerCreated(mongoDBCAPI.libmongodbcapi_client_create(instance, status));
         } catch (Throwable t) {
-            throw createError("db_client_new", t);
+            throw createError("client_create", t);
         }
     }
 
-    static void db_client_destroy(final Pointer client) {
+    static void client_destroy(final Pointer client) {
         checkInitialized();
         try {
-            mongoDBCAPI.libmongodbcapi_db_client_destroy(client);
+            validateErrorCode(mongoDBCAPI.libmongodbcapi_client_destroy(client, status));
         } catch (Throwable t) {
-            throw createError("db_client_destroy", t);
+            throw createError("client_destroy", t);
         }
     }
 
-    static void db_client_wire_protocol_rpc(final Pointer client, final byte[] input, final int inputSize,
-                                                         final PointerByReference output, final IntByReference outputSize) {
+    static void client_invoke(final Pointer client, final byte[] input, final PointerByReference output, final IntByReference outputSize) {
         checkInitialized();
         try {
-            validateErrorCode(mongoDBCAPI.libmongodbcapi_db_client_wire_protocol_rpc(client, input, inputSize, output, outputSize));
+            validateErrorCode(mongoDBCAPI.libmongodbcapi_client_invoke(client, input, input.length, output, outputSize, status));
         } catch (Throwable t) {
-            throw createError("db_client_wire_protocol_rpc", t);
+            throw createError("client_invoke", t);
```

Comment:
validateErrorCode can throw MongoClientException, which is then immediately wrapped in another MongoClientException.  Can we special case that in an extra try block so we avoid that wrapping (which just creates more confusing stack traces).

Reply:
Done.